archive/100-theorems-list/81_sum_of_prime_reciprocals_diverges.lean

```diff
@@ -0,0 +1,220 @@
+/-
+Copyright (c) 2021 Manuel Candales. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Manuel Candales
+-/
+import topology.instances.ennreal
+import algebra.squarefree
+
+/-!
+# Divergence of the Prime Reciprocal Series
+
+This file proves Theorem 81 from the [100 Theorems List](https://www.cs.ru.nl/~freek/100/).
+The theorem states that the sum of the reciprocals of all prime numbers diverges.
+The formalization follows Erdős's proof by upper and lower estimates.
+
+## References
+
+https://en.wikipedia.org/wiki/Divergence_of_the_sum_of_the_reciprocals_of_the_primes
+-/
+
+open_locale big_operators
+open_locale classical
+open filter finset
+
+lemma card_le_div_nat {n p : ℕ} (hp : 0 < p) : card {e ∈ range n | p ∣ e + 1} ≤ n / p :=
+begin
+  let f : ℕ → ℕ := λ e, (e + 1) / p - 1,
+  let Np := {e ∈ range n | p ∣ e + 1},
+
+  have hf : ∀ a : ℕ, a ∈ Np → f a ∈ range (n / p),
+  { simp only [f, Np, sep_def, mem_filter, mem_range],
+    rintros a ⟨han, ⟨w, hw⟩⟩,
+    have hnp : 1 ≤ n / p,
+    { have : 0 < w := by nlinarith,
+      rw ← nat.div_self hp,
+      exact nat.div_le_div_right (by nlinarith) },
+    calc (a + 1) / p - 1 ≤ n / p - 1 : nat.sub_le_sub_right
+                                         (nat.div_le_div_right (nat.succ_le_iff.mpr han)) 1
+    ...                  < n / p     : nat.sub_lt (zero_lt_one.trans_le hnp) zero_lt_one },
+
+  have hf_inj : ∀ a₁ : ℕ, a₁ ∈ Np → ∀ a₂ : ℕ, a₂ ∈ Np → f a₁ = f a₂ → a₁ = a₂,
+  { simp_intros a₁ ha₁ a₂ ha₂ hfeq only [f, Np, sep_def, mem_filter, mem_range],
+    obtain ⟨⟨hna₁, ⟨w₁, hw₁⟩⟩, ⟨hna₂, ⟨w₂, hw₂⟩⟩⟩ := ⟨ha₁, ha₂⟩,
+    rw [hw₁, hw₂, nat.mul_div_cancel_left w₁ hp, nat.mul_div_cancel_left w₂ hp] at hfeq,
+    have hw₁_eq_w₂ : w₁ = w₂, { refine nat.pred_inj _ _ hfeq; nlinarith },
+    rw [← add_left_inj 1, hw₁, hw₂, hw₁_eq_w₂] },
+
+  calc  card Np ≤ card (range (n / p)) : card_le_card_of_inj_on f hf hf_inj
+  ...           = n / p                : card_range (n / p),
+end
+
+lemma card_le_div_real {n p : ℕ} (hp : 0 < p) :
+  (card {e ∈ range n | p ∣ e + 1} : ℝ) ≤ n / p :=
+calc  (card {e ∈ range n | p ∣ e + 1} : ℝ)
+    ≤ ↑(n / p) : nat.cast_le.mpr $ card_le_div_nat hp
+... ≤ n / p    : (le_div_iff $ by exact_mod_cast hp).mpr $ by apply_mod_cast nat.div_mul_le_self
+
+lemma card_eq_card_sdiff_add_card {A B : finset ℕ} (h : A ⊆ B) :
```

Comment:
Please add this as general lemma (so for `finset \alpha`) to the file where `finset.card` is defined.